{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This notebook contains code that can be used to create a custom taxonomy from a phylogenetic tree that can be used for Kraken or emboss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the phylogenetic tree into memory\n",
    "from ete3 import Tree\n",
    "from os import chdir, path, makedirs, system\n",
    "from Bio.SeqIO import index as fasta_index,write\n",
    "\n",
    "chdir(\"/mnt/research/germs/shane/hgt/data\")\n",
    "prot=\"Cas1_98_AVX2\"\n",
    "phyloTree = Tree(\"trees/RAxML_result.\"+prot)\n",
    "chdir(\"/mnt/research/germs/shane/antibioticResistance/data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate the custom taxonomy for Kraken based on the loaded phylogenetic tree\n",
    "specificDB = \"dbs/\"+prot.replace('C','c')+\"DB/taxonomy\"\n",
    "if not path.exists(specificDB): makedirs(specificDB);print(\"Made \"+specificDB)\n",
    "nodesTracker={}\n",
    "names= open(specificDB+\"/names.dmp\",'w')\n",
    "nodes= open(specificDB+\"/nodes.dmp\",'w')\n",
    "root = phyloTree.get_tree_root()\n",
    "\n",
    "for taxID,node in enumerate(phyloTree.traverse(\"preorder\")):\n",
    "    nodesTracker[node]=str(taxID+1)\n",
    "    if not node.is_leaf():\n",
    "        node.name = str(taxID+1)\n",
    "        if node ==root:nodes.write(\"\\t|\\t\".join([str(taxID+1),'1',\"no rank\",\"-\"])+'\\n') # Root is its own parent\n",
    "        else: nodes.write(\"\\t|\\t\".join([str(taxID+1),nodesTracker[node.up],\"no rank\",\"-\"])+'\\n')\n",
    "    else: nodes.write(\"\\t|\\t\".join([str(taxID+1),nodesTracker[node.up],\"species\",\"-\"])+'\\n')\n",
    "    names.write(\"\\t|\\t\".join([str(taxID+1),node.name,\"-\",\"scientific name\",\"\"])+'\\n')\n",
    "print(len(phyloTree))\n",
    "names.close()\n",
    "nodes.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rewrite the fasta file with the taxonomy IDs generated in Cell 2 so that Kraken can link taxonomy to the sequences\n",
    "allSeqs = fasta_index(\"assemblies/Cas1_Class2_Coding.fa\",\"fasta\")\n",
    "missing,good=0,0\n",
    "with open(\"assemblies/Cas1-CodingSeqsCleaned.fa\",'w') as fh:\n",
    "    for leaf in phyloTree.get_leaves():\n",
    "        if leaf.name not in allSeqs: missing+=1;continue\n",
    "        rec=allSeqs[leaf.name]\n",
    "        rec.id = rec.id + \"|kraken:taxid|\" + nodesTracker[leaf]+ \" Cas12 ORF Region\"\n",
    "        rec.name = rec.id\n",
    "        rec.description = \"\"\n",
    "        write(rec,fh,\"fasta\")\n",
    "        good+=1\n",
    "missing,good"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "#Add the sequences that with the added taxIDs generated by this module to a custom database\n",
    "kraken-build --add-to-library assemblies/Cas1-CodingCleaned.fa --db dbs/cas1_98_AVX2DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#Build the Kraken database\n",
    "kraken-build --threads 15 --jellyfish-hash-size 19200M --max-db-size 100 --build --db dbs/cas1_98_AVX2DB"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
